crypto/x509/by_store.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/store.h>
+#include "internal/cryptlib.h"
+#include "internal/x509_int.h"
+#include "x509_lcl.h"
+
+/* Generic object loader, given expected type and criterion */
+static int cache_objects(X509_LOOKUP *lctx, const char *uri,
+                         X509_LOOKUP_TYPE type,
+                         const OSSL_STORE_SEARCH *criterion,
+                         int depth)
+{
+    int ok = 0;
+    OSSL_STORE_CTX *ctx = NULL;
+    int expect = 0;
+    X509_STORE *xstore = X509_LOOKUP_get_store(lctx);
+
+    if (type == X509_LU_X509)
+        expect = OSSL_STORE_INFO_CERT;
+    else if (type == X509_LU_X509)
+        expect = OSSL_STORE_INFO_CRL;
+    if (expect == 0) {
+        X509err(X509_F_CACHE_OBJECTS, X509_R_WRONG_LOOKUP_TYPE);
+        return 0;
+    }
+
+    if ((ctx = OSSL_STORE_open(uri, NULL, NULL, NULL, NULL)) == NULL)
+        return 0;
+    OSSL_STORE_expect(ctx, expect);
+    if (criterion != NULL)
+        OSSL_STORE_find(ctx, criterion);
+
+    while (!OSSL_STORE_eof(ctx) && !OSSL_STORE_error(ctx)) {
+        OSSL_STORE_INFO *info = OSSL_STORE_load(ctx);
+        int infotype = info == 0 ? 0 : OSSL_STORE_INFO_get_type(info);
+
+        ok = 0;
+
+        if (info == NULL)
```

Comment:
Why do not break similar as code in store util? There is no need to test for eof and error if load returns null.

Reply:
Good point